### PR TITLE
Error Boundary Handling

### PR DESCRIPTION
**Pull request by spaceo**
#### Link to issue

https://reload.atlassian.net/browse/DDFSOEG-246

#### Description

We need a "catch all" way of treating errors that happens while fetching.
This PR makes sure that errors caused by fetching will be catched. Depending on there severity the React error boundary will be invoked.

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Notes
There is a PatronPage Cypress test failing, because one or more fixtures are missing.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -146,6 +147,7 @@
     "focus-trap-react": "^10.0.2",
     "graphql": "^16.5.0",
     "graphql-tag": "^2.12.6",
+    "http-error-classes": "^1.0.1",
```
Comment (kasperg) [inline package.json:150]: I do not understand this choice of library:

1. As I see it we do not use it for anything special apart from a single error type with some predefined fields. 
2. We do not use all the predefined error classes. Any they do not have much relevance to us.
3. It is not very widely used and/or from a well-known community actor.

Based on your work with `FetcherError` it seems we could have done this ourselves or used [a more popular package for managing custom error types](https://www.npmjs.com/search?q=custom%20errors).

**Review by jeppekroghitk — APPROVED** [review]

File: src/components/store.jsx [hunk 2]
```diff
@@ -4,17 +4,40 @@ import { Provider } from "react-redux";
...
 const queryClient = new QueryClient({
   defaultOptions: {
     queries: {
       // We need to have some sort of client caching strategy.
       // This is for temporarily testing.
       // In this case cache needs be updated after 30 seconds:
-      staleTime: 1000 * 30
+      staleTime: 1000 * 30,
+      // useErrorBoundary: true
+      useErrorBoundary: queryErrorHandler
```
[the remark this answers is not in the document]
Comment (spaceo) [reply src/components/store.jsx:33]: Shouldn't be there...

File: src/core/cover-service-api/mutator/fetcher.ts [hunk 3]
```diff
@@ -39,28 +33,41 @@ export const fetcher = async <ResponseType>({
...
+  } catch (error: unknown) {
+    if (error instanceof CoverServiceHttpError) {
+      throw error;
+    }
+    const message = error instanceof Error ? error.message : "Unknown error";
+    throw new FetchFailedError(message, serviceUrl);
+  }
   // Do nothing. Some of our responses are intentionally empty and thus
   // cannot be converted to JSON. Fetch API and TypeScript has no clean
   // way for us to identify empty responses so instead we swallow
```
[the remark this answers is not in the document]
Comment (spaceo) [reply src/core/cover-service-api/mutator/fetcher.ts:73]: No. The comment belongs to the line returning `null` (doing nothing). So I keep it there.
Would be nice with one Orval fetcher though :)

File: src/core/cover-service-api/mutator/fetcher.ts [hunk 4]
```diff
@@ -39,28 +33,41 @@ export const fetcher = async <ResponseType>({
...
+      if (!(e instanceof SyntaxError)) {
+        throw e;
+      }
+    }
+  } catch (error: unknown) {
+    if (error instanceof CoverServiceHttpError) {
+      throw error;
+    }
+    const message = error instanceof Error ? error.message : "Unknown error";
+    throw new FetchFailedError(message, serviceUrl);
```
[the remark this answers is not in the document]
Comment (spaceo) [reply src/core/cover-service-api/mutator/fetcher.ts:69]: TI'll rather not touch it. 
`export type ErrorType<ErrorData> = ErrorData & { status:` is connected to the generated code outside of the fetcher which expects it to be exported.
When writing one common Orval fetcher, we should see if we could consolidate it somehow.

File: src/core/fetchers/FetcherError.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+export class FetcherError<ContextType> extends Error {
```
[the remark this answers is not in the document]
Comment (spaceo) [reply src/core/fetchers/FetcherError.ts:1]: I'll look into it

File: docs/error_handling.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Error Handling
+
+Error handling is something that is done on multiple levels:
+Eg.: Form validation, network/fetch error handling, runtime errors.
+You could also argue that the fact that the codebase is making use of typescript
+and code generation from the various http services (graphql/REST) belongs to
+the same idiom of error handling in order to make the applications more robust.
+
```
[the remark this answers is not in the document]
Comment (spaceo) [reply docs/error_handling.md:8]:
```quote
We did not. We also chose to add a name property which is never used.
```
That is not correct. By setting *.name on an error class we mange what the name is. And we are checking the name of the classes [here](https://github.com/reload/dpl-react/blob/cec3d3b3e9d9f44f7204ba08949fc66096e8a424/src/components/store.jsx#L14-L16).

[the remark this answers is not in the document]
Comment (spaceo) [reply docs/error_handling.md:8]: The whole concept of cirtical and not critical is already described in the documentation. Don't know what I should add more?

File: src/core/dpl-cms/mutator/fetcher.ts [hunk 7]
```diff
@@ -1,44 +1,11 @@
...
-          params.append(key, inner.toString());
-        });
-      } else {
-        params.append(key, value.toString());
-      }
-    });
-  }
-
-  return params;
-}
```
[the remark this answers is not in the document]
Comment (kasperg) [reply src/core/dpl-cms/mutator/fetcher.ts:41]: Ah - I had not seed that the function was already defined there and thus not added as a part of this PR.

File: src/core/cover-service-api/mutator/fetcher.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kasperg) [reply src/core/cover-service-api/mutator/fetcher.ts:73]: I agree about the two belonging together. How about moving the comment *and* the final `return null`?

The following `catch` on L64 will always `throw` anyhow so the only (quite convoluted) way to end up with the final `return null` in L75 is by not `catch`ing a `SyntaxError` in L59.

File: package.json (hunk 1 above)
Comment (spaceo) [reply package.json]: You're right. I have replaced the class with our own error class. The key concepts of the class is the status code and the possibility of giving it a context.

File: src/core/dbc-gateway/DbcGateWayHttpError.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import FetcherCriticalHttpError from "../fetchers/FetcherCriticalHttpError";
+
+export default class DbcGateWayHttpError<
+  ContextType
+> extends FetcherCriticalHttpError<ContextType> {
```
[the remark this answers is not in the document]
Comment (spaceo) [reply src/core/dbc-gateway/DbcGateWayHttpError.ts]: done (also with the other critical ones)

File: src/core/publizon/mutator/PublizonServiceHttpError.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import FetcherCriticalHttpError from "../../fetchers/FetcherCriticalHttpError";
+
+// This HttpError is critical because we want to inform the user that something went wrong
+// while ordering a digitial article.
+export default class PublizonServiceHttpError<
+  ContextType
```
Comment (kasperg) [inline src/core/publizon/mutator/PublizonServiceHttpError.ts:6]: One of us is confused here :smile:. As far as I know Publizon is not related to ordering of digital articles - at the moment at least.

Publizon has to do with checking the availability of digital materials and as such I do not see it as critical.

**Comment by kasperg** [comment]
The failing test blocks merging of this. I suggest that we [`skip it`](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Excluding-and-Including-Tests) and add a `TODO`.

**Review by kasperg — APPROVED** [review]
:+1: LGTM

**Pull request merged**